### ISSUE TITLE
FT-2 — Pseudolabel stage, fine-tune presets/defaults, and CI-tested fine-tuning docs

### ISSUE THREAD

**Issue by aacostadiaz** [ACEsuit/mace#1581]
**Depends on:** FT-1 (#1580), CFG-1 (#1574) · **Blocks:** —

**Prerequisite:** the Phase 0 characterization suite is committed — goldens, fixtures and the single-source tolerance table under `tests/golden/`, characterization tests under `tests/unit/`. References of the form `P0-N` name that work.

**Context:** Legacy pseudolabel replay (`--pseudolabel_replay`, `mace/tools/arg_parser.py:546-551`) has four defects, all fixed here. (1) **Silent partial failure**: a function-level `except Exception` returns `False` with training continuing on a warning (`mace/tools/multihead_tools.py:509-511`, `mace/cli/run_train.py:645-656`) — a replay set can contain no pseudolabels at all and the run carries on. The batch-level handler is no longer part of this: it raises rather than substituting the file's own labels for a failed batch (`multihead_tools.py:380-392`), so a set can no longer mix FM pseudolabels with DFT labels from a different level of theory. Defects (2) to (4) below are unchanged and still present. (2) **Every DDP rank regenerates independently**: the call at `run_train.py:646-659` has no rank guard — cost ×`world_size`, and fp32 GPU `scatter_add_` nondeterminism means ranks can hold different labels for the same configuration. (3) **No provenance**: labels depend on (FM weights, config set, device, dtype, torch version, batch partitioning) and none of it is recorded; nothing persists the labels at all. (4) **`requires_grad` juggling** (`multihead_tools.py:255-259` off, `:351-353` restore) instead of `torch.no_grad()` — generation welded into training rather than a callable stage. Separately, multihead fine-tuning mutates the args namespace — `lr=0.0001`, `ema=True`, `ema_decay=0.99999` unless `--force_mh_ft_lr` (`run_train.py:243-249`, flag at `arg_parser.py:597-601`) — the exact pattern the v1 config system kills: presets are declared defaults, never post-parse mutations, and there is no `force_mh_ft_lr` equivalent in v1. This ticket builds the v1 pseudolabel stage on FT-1 (#1580)'s uniform head interface, the fine-tune shortcuts/presets (auto-download curated replay sets, replay-dependent defaults), and worked, CI-tested fine-tuning docs.

**Interface & constraints:**

1. **Caching stance: no implicit cross-run cache.** Pseudolabels are generated **once per training run, on rank 0 only**, and persisted into the run directory as a first-class artifact. Rationale: generation ≈ 15 min vs fits of 2–24 h (the 15-min figure is ~10–20× pessimistic: 44–85 s on one A100, i.e. 0.05–1.2 % of run time), and GPU generation is not bit-reproducible at fp32 **or fp64**, so a content-addressed cache would record a *promise* while the run artifact records a *fact*. Reuse across runs is always **explicit and validated**.

2. **Locations.** Downloaded replay *inputs* stay in `~/.cache/mace` (true cache: immutable upstream; `get_cache_dir`, `mace/tools/utils.py:170-172`). Generated labels + provenance go to `<run_dir>/pseudolabels/<head>/` — never to `~/.cache`, never next to the shared dataset, and never as the working-directory scratch XYZ legacy leaves behind (`multihead_tools.py:182`). Explicitly reused labels live wherever the user's config points, validated at load.

3. **Config** (`mace_core.config.finetune`, in CFG-1 (#1574)'s schema):

```python
class PseudolabelSpec(BaseModel):
    enabled: bool = False                                  # ex --pseudolabel_replay
    properties: frozenset[Literal["energy", "forces", "stress", "virials", "dipole", "charges"]] = \
        frozenset({"energy", "forces"})                    # legacy writes all six when present,
                                                           # multihead_tools.py:302-341
    stress_if_missing: bool = False                        # ex --pseudolabel_replay_compute_stress
                                                           # (legacy gating: multihead_tools.py:296-321)
    batch_size: int | None = None                          # None -> training batch size
    labels_from: Path | None = None                        # EXPLICIT reuse of a prior artifact;
                                                           # mutually exclusive with enabled=True (validator, CFG-1 (#1574))
```

4. **Generation is a callable stage** (`mace_torch.finetune.pseudolabels`) — pure w.r.t. inputs, under `torch.no_grad()`; **any per-batch failure raises** (the keep-original-labels fallback is a recorded DROP):

```python
class PseudolabelProvenance(BaseModel):
    foundation_model: str                 # registry id and/or sha256 of the FM weights payload
    dataset_fingerprint: str              # canonical content hash of the INPUT configs (DATA-1 (#1569)'s fingerprint)
    spec: PseudolabelSpec
    device: str; dtype: str
    torch_version: str; kernel_backend: str
    generated_at: datetime
    n_configs: int
    deterministic: bool                   # True iff CPU or deterministic segment_reduce

def generate_pseudolabels(model, configs, spec, *, device) -> tuple[list[Configuration], PseudolabelProvenance]: ...
def write_pseudolabel_artifact(run_dir, head, configs, provenance) -> Path: ...
def load_pseudolabel_artifact(path) -> tuple[list[Configuration], PseudolabelProvenance]: ...

# Artifact format: a binary shard that preserves fp64 labels exactly — the same
# shard writer and typed JSON manifest used for prepared datasets — with the
# PseudolabelProvenance sidecar as JSON beside it. A text format such as extxyz
# is NOT acceptable: exact round-trip and byte-identical labels across ranks both
# require lossless numeric storage.
def validate_pseudolabel_reuse(prov, *, current_fm_fingerprint, current_dataset_fingerprint) -> None: ...
```

Property weights travel with the labels. Legacy already carries this subtlety and it must be preserved: a stress pseudolabel written because the original config *had* stress keeps the original `property_weights["stress"]`, while one written only because `stress_if_missing` was set gets weight 1.0 (`multihead_tools.py:296-321`). Getting that backwards silently reweights the replay loss.

One implementation, two entry points: the same `generate_pseudolabels` backs the in-run path and the standalone `mace data pseudolabel` subcommand (registered under CLI-1 (#1579)'s `mace data` group), which emits the identical artifact — sweeps generate once and point N configs at it via `labels_from`.

5. **Reuse validation** (`validate_pseudolabel_reuse`): FM-fingerprint mismatch is a **hard error** — no override in v1 (relabeling costs ~15 min); dataset-fingerprint mismatch is a hard error; `torch_version`/`kernel_backend` drift with `deterministic=False` warns (labels valid but not re-derivable).

6. **DDP contract** (TRN-4 (#1578) provides the primitives; this ticket owns the contract): rank 0 generates and writes the artifact; `barrier()`; all ranks load it — byte-identical labels everywhere, **one label set per run**. Where `run_dir` is not on a shared filesystem, rank 0 broadcasts the serialized artifact via `broadcast_object_list`.

7. **Recorded DROPs vs legacy** (traceability rows in P0-0's matrix, pinned by P0-5 characterization tests before this path replaces them): the batch-failure keep-original fallback (`multihead_tools.py:343-349`); the function-level `except → return False → continue with a warning` (`:452-454`, `run_train.py:658-659`); per-rank generation; and the `no_data_ok` unlabelled-continuation (`multihead_tools.py:214-218`, `run_train.py:381-385`, warning emitted at `mace/data/utils.py:300-301`), which today lets a replay head start with *zero* configurations as long as pseudolabelling is enabled.

8. **Shortcuts/presets and replay-dependent defaults.** Best-practice presets cover: source of configs (auto-download curated replay sets vs file — the named sets are `mp`, `omat`, `matpes_pbe`, `matpes_r2scan`, today's four fixed URLs at `multihead_tools.py:154-165`, downloaded via FT-1 (#1580)'s replay assembly), source of labels (pseudolabels), subselection strategy and count. The replay preset sets `PseudolabelSpec(enabled=True, properties={energy, forces})` for the FM-shortcut replay sets. Defaults that legacy forced by mutating args (lr, EMA) become **replay-dependent defaults resolved by config parsing**; explicit user settings always override; there is no `force_mh_ft_lr` equivalent, and the resolved config records which preset supplied each value.

9. **Replay-head E0s.** Pseudolabel energies are consistent with FM E0s by construction (the FM produces them); the implicit replay head's E0 spec is fixed to `E0sFromFoundation(missing="error")`, reading values from the FM artifact's `e0s` metadata block — never from `model_foundation.atomic_energies_fn.atomic_energies` by reflection as legacy does (`run_train.py:565-576`), and never with the head-0 auto-pick legacy falls back to for a multi-head FM (`:563-565`, which only logs). The missing-element policy (`missing = "error" (default) / "average" (lstsq restricted to unresolved elements, FM values held fixed) / "zero" (explicit, validated at data load)`) replaces today's inconsistent failure modes; re-estimation writes a *new* `e0s` block — **FM metadata is immutable**. The new-species parameter-initialization side is FT-3 (#1582).

10. **Docs.** Worked fine-tuning examples (replay presets, pseudolabels, explicit reuse across a sweep, standalone `mace data pseudolabel`) with expected outputs shown, executed in CI so they cannot rot.

**Task:**
1. `mace_core.config.finetune`: `PseudolabelSpec` + `PseudolabelProvenance` schemas; the `labels_from ⊕ enabled` validator lands with CFG-1 (#1574)'s cross-section validation.
2. `mace_torch.finetune.pseudolabels`: the generate/write/load/validate quartet — `torch.no_grad()`, hard-fail on any per-batch failure, artifact at `<run_dir>/pseudolabels/<head>/` with the provenance sidecar, and the two-valued stress property weight preserved.
3. Rank-0-generate / barrier / all-ranks-load wiring in the training pipeline (TRN-4 (#1578) primitives), with the `broadcast_object_list` fallback.
4. `mace data pseudolabel` subcommand emitting the identical artifact.
5. Fine-tune shortcuts/presets: named curated replay sets (`mp`, `omat`, `matpes_pbe`, `matpes_r2scan`) via FT-1 (#1580)'s assembly; replay-dependent lr/EMA defaults as declared config defaults with user override; no `force_mh_ft_lr` equivalent.
6. Replay-head E0 policy: `E0sFromFoundation(missing="error")` default, the three-valued `missing` policy, explicit head selection for a multi-head FM, FM-metadata immutability on re-estimation.
7. Worked fine-tuning docs wired into CI execution.

**Out of scope:** replay data assembly, FPS subselection, LoRA/freezing (FT-1 (#1580)); the DDP machinery itself (TRN-4 (#1578)); the `E0Spec` union and validators as such (CFG-1 (#1574)) and new-species parameter initialization (FT-3 (#1582)); the MDP fine-tune mode (FT-4 (#1637)); dataset fingerprint implementation (DATA-1 (#1569)); any cross-run pseudolabel cache (rejected by decision — do not add one "for convenience").

**Acceptance criteria:**
- [ ] An injected per-batch inference failure raises and aborts the run — no silent mixing of FM pseudolabels with original labels (both legacy fallbacks demonstrably gone).
- [ ] A replay head that resolves to zero configurations is a hard error; the `no_data_ok` continuation has no v1 counterpart.
- [ ] `generate_pseudolabels` runs under `torch.no_grad()` (no `requires_grad` save/restore) and is pure w.r.t. inputs; the artifact round-trips: `write_pseudolabel_artifact` → `load_pseudolabel_artifact` reproduces configs + provenance exactly, including per-property weights.
- [ ] A stress label written because the source config had stress keeps the source weight; one written only because `stress_if_missing` was set gets weight 1.0.
- [ ] Provenance records FM fingerprint, dataset fingerprint, spec, device/dtype, torch/backend versions, `deterministic` flag; labels are written under `<run_dir>/pseudolabels/<head>/` and never under `~/.cache/mace` or the working directory.
- [ ] `labels_from` with a mismatched FM fingerprint or dataset fingerprint is a hard error; version drift with `deterministic=False` warns; `labels_from` + `enabled=True` fails config validation.
- [ ] CPU-DDP smoke (gloo, 2 processes): rank 0 generates, all ranks load byte-identical labels — one label set per run.
- [ ] `mace data pseudolabel` emits an artifact schema-identical to the in-run path, and a training config consuming it via `labels_from` trains.
- [ ] The replay preset resolves to `enabled=True, properties={energy, forces}` on the curated sets; replay-dependent lr/EMA defaults appear in the resolved config as declared defaults; an explicit user setting overrides every preset value; no `force_mh_ft_lr`-style flag exists and no args-namespace mutation remains.
- [ ] Replay head E0s resolve as `E0sFromFoundation(missing="error")` from artifact metadata; a multi-head FM without an explicit head is an error, not a head-0 pick; `missing="zero"` is validated at data load; re-estimation leaves FM metadata untouched.
- [ ] The worked fine-tuning docs execute in CI and their shown expected outputs match.

**Verify:**
```bash
python -m pytest packages/mace-torch/tests/finetune/test_pseudolabels.py -v   # stage purity, hard-fail, artifact round-trip, reuse validation
python -m pytest packages/mace-torch/tests/finetune/test_pseudolabel_ddp.py -v # gloo 2-proc rank-0 contract, byte-identical labels
python -m pytest packages/mace-torch/tests/finetune/test_presets.py -v        # replay-dependent defaults, override precedence, no silent mutation
python -m pytest docs/finetuning -v                                            # worked examples execute with shown outputs
```
**Review focus:** the failure semantics — every legacy keep-going path (batch fallback, function-level `except`, unlabelled `no_data_ok` continuation) must be a loud error with its DROP pinned; the two-valued stress weight; and that reuse validation cannot be bypassed (no `--force`-style override slipped in).

---
